java/code/src/com/suse/manager/maintenance/IcalUtils.java

```diff
@@ -0,0 +1,211 @@
+     *
+     * The algorithm only checks maintenance windows within roughly a year and a month since the startDate.
+     *
+     * @param calendar the {@link Calendar}
+     * @param eventName for MULTI calendars: only deal with events with this name, filter out the rest
+     * @param startDate the start date
+     * @param limit upper limit of maintenance windows to return
+     * @return the list of upcoming maintenance windows
+     */
+    public Stream<Pair<Instant, Instant>> calculateUpcomingPeriods(Calendar calendar, Optional<String> eventName,
+            Instant startDate, int limit) {
+        ComponentList<CalendarComponent> allEvents = calendar.getComponents(Component.VEVENT);
+
+        Collection<CalendarComponent> filteredEvents = eventName.map(name -> {
+            Predicate<CalendarComponent> summary = c -> c.getProperty("SUMMARY").equals(name);
+            Predicate<CalendarComponent>[] ps = new Predicate[]{summary};
+            Filter<CalendarComponent> filter = new Filter<>(ps, Filter.MATCH_ALL);
+            return filter.filter(allEvents);
+        }).orElse(allEvents);
+
+        // we will look a year and month to the future
+        Period period = new Period(new DateTime(startDate.toEpochMilli()), Duration.ofDays(365 + 31));
+
+        List<PeriodList> periodLists = filteredEvents.stream()
+                .map(c -> c.calculateRecurrenceSet(period))
+                .filter(l -> !l.isEmpty())
+                .collect(toList());
+
+        Stream<Pair<Instant, Instant>> sortedLimited = periodLists.stream()
+                .map(pl -> pl.stream())
+                .reduce(Stream.empty(), Stream::concat)
+                .sorted()
+                .limit(limit)
+                .map(p -> Pair.of(p.getStart().toInstant(), p.getRangeEnd().toInstant()));
+
+        return sortedLimited;
+    }
+
+    /**
+     * Get all schedules of given calendar at given date.
+     * Filter results by summary, if the summary parameter is passed.
+     *
+     * @param date the date
+     * @param calendar the calendar
+     * @param summary event summary
+     * @return the collection of calendars components matching given date (and optionally summary)
+     */
+    public Collection<CalendarComponent> getCalendarEventsAtDate(Date date, Optional<Calendar> calendar,
+            Optional<String> summary) {
+        if (calendar.isEmpty()) {
+            return emptySet();
+        }
+
+        Period p = new Period(new DateTime(date), java.time.Duration.ofSeconds(1));
+        ArrayList<Predicate<Component>> rules = new ArrayList<>();
+        rules.add(new PeriodRule<>(p));
+
+        summary.ifPresent(s -> {
+            Summary filterSummary = new Summary(s);
+            HasPropertyRule<Component> propertyRule = new HasPropertyRule<>(filterSummary);
+            rules.add(propertyRule);
+        });
+
+        @SuppressWarnings("unchecked")
```

Comment:
I guess that one slipped through the net ;)

Reply:
Uf, I actually didn't want to remove these annotations in this PR (it's a separate task in our task list). Removing the others was "an accident". I'll leave it like this for now, but we need to talk about it with @mcalmer , who said, it produced some errors in his IDE.

Reply:
Right let's have a discussion about this. This can also be addressed later